offchain-modules/packages/x/src/ckb/tx-helper/generator.ts

```diff
@@ -613,13 +613,184 @@ export class CkbTxGenerator extends CkbTxHelper {
+              index: ForceBridgeCore.config.ckb.deps.pwLock!.cellDep.outPoint.index,
+            },
+            dep_type: ForceBridgeCore.config.ckb.deps.pwLock!.cellDep.depType,
+          },
+          {
+            out_point: {
+              tx_hash: secp256k1.TX_HASH,
+              index: secp256k1.INDEX,
+            },
+            dep_type: secp256k1.DEP_TYPE,
+          },
+        );
+      });
+    } else {
+      throw new Error('unsupported sender address!');
+    }
+
+    const sudtScript = {
+      code_hash: ForceBridgeCore.config.ckb.deps.sudtType.script.codeHash,
+      hash_type: ForceBridgeCore.config.ckb.deps.sudtType.script.hashType,
+      args: sudtArgs,
+    };
+    const bridgeFee = BigInt(ForceBridgeCore.config.eth.lockNervosAssetFee);
+
+    // add committee multisig output
+    const committeeMultisigLockscript = parseAddress(getOmniLockMultisigAddress());
+    const committeeMultisigCell: Cell = {
+      cell_output: {
+        capacity: `0x${bridgeFee.toString(16)}`, // bridge fee
+        lock: committeeMultisigLockscript,
+        type: sudtScript,
+      },
+      data: utils.toBigUInt128LE(amount), // locked sudt amount
+    };
+    txSkeleton = txSkeleton.update('outputs', (outputs) => {
+      return outputs.push(committeeMultisigCell);
+    });
+
+    // collect sudt
+    const searchKey = {
+      script: senderLockscript,
+      script_type: ScriptType.lock,
+      filter: {
+        script: sudtScript,
+      },
+    };
+    const sudtCells = await this.collector.collectSudtByAmount(searchKey, amount);
+    const total = sudtCells.map((cell) => utils.readBigUInt128LE(cell.data)).reduce((a, b) => a + b, 0n);
+    if (total < amount) {
+      throw new Error('sudt amount is not enough!');
+    }
+    logger.debug('lock sudtCells: ', sudtCells);
+
+    // ckb change output cell
+    const senderOutputCkbChangeCell: Cell = {
+      cell_output: {
+        capacity: '0x0',
+        lock: senderLockscript,
+      },
+      data: '0x',
+    };
+    const senderOutputCkbChangeCellMinimalCapacity = minimalCellCapacity(senderOutputCkbChangeCell);
+    let senderOutputSudtChangeCellMinimalCapacity = BigInt(0);
+    if (total > amount) {
+      // sudt change output cell
+      const senderOutputSudtChangeCell: Cell = {
+        cell_output: {
+          capacity: '0x0',
+          lock: senderLockscript,
+          type: sudtScript,
+        },
+        data: utils.toBigUInt128LE(total - amount),
+      };
+      senderOutputSudtChangeCellMinimalCapacity = minimalCellCapacity(senderOutputSudtChangeCell);
+      senderOutputSudtChangeCell.cell_output.capacity = `0x${senderOutputSudtChangeCellMinimalCapacity.toString(16)}`;
+      // add sudt change output
+      txSkeleton = txSkeleton.update('outputs', (outputs) => {
+        return outputs.push(senderOutputSudtChangeCell);
+      });
+    }
+    // collect ckb
+    const senderNeedCkbSupplyCapacity =
```

Comment:
Calculate logic is wrong for `senderNeedCkbSupplyCapacity`. 
```
const outputNeedCapacity = bridgeFee + senderOutputSudtChangeCellMinimalCapacity + senderOutputCkbChangeCellMinimalCapacity + 1n * BigInt(100000000)`;

if (totalCapacityOfSudtCellsInInputs >= outputNeedCapacity) {
  senderNeedCkbSupplyCapacity = 0; 
  // next: refund change capacity to senderOutputCkbChangeCell
} else {
  senderNeedCkbSupplyCapacity = outputNeedCapacity - totalCapacityOfSudtCellsInInputs;
  // next: collect capacity cells of user, then refund change capacity to senderOutputCkbChangeCell
}


```

Reply:
fixed this logic, collect ckb when inputs capacity does not offset outputs capacity, and to refund directly when the input capacity is sufficient